dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java

```diff
@@ -1095,9 +1096,44 @@ public void endVisit(Initializer x, MethodScope scope) {
     public void endVisit(InstanceOfExpression x, BlockScope scope) {
       try {
         SourceInfo info = makeSourceInfo(x);
-        JExpression expr = pop(x.expression);
+        JExpression expr;
+        JDeclarationStatement jDeclarationStatement = null;
+        if (x.pattern != null) {
+          jDeclarationStatement = (JDeclarationStatement) pop();
+        }
+        expr = pop(x.expression);
         JReferenceType testType = (JReferenceType) typeMap.get(x.type.resolvedType);
-        push(new JInstanceOf(info, testType, expr));
+
+        if (jDeclarationStatement == null) {
+          push(new JInstanceOf(info, testType, expr));
+        } else {
+          String localVariableName = jDeclarationStatement.getVariableRef().getTarget().getName();
+          if (!curMethod.instanceOfDeclarations.containsKey(localVariableName)) {
+            curMethod.body.getBlock().addStmt(0, jDeclarationStatement);
+            curMethod.instanceOfDeclarations.put(localVariableName, jDeclarationStatement);
```

Comment:
Admittedly I haven't taken the time to dig in and try this, but I see that `x.pattern` is a jdt `Pattern`, which has a (public?) `secretPatternVariable` field of type `LocalVariableBinding` - does it work to use that in the existing `MethodInfo.locals` map instead of making a new `instanceOfDeclarations` map?

Reply:
Just to add a clarification on the case we are trying to solve here, in this sample

```java
    Shape shape1 = new Square();
    Shape shape2 = new Square();

    if (shape1 instanceof Square square && square.getLength() > 0) {
    }
    if (shape2 instanceof Square square && square.getLength() > 0) {
    }
```

According to our solution we will be rewriting this to 

```java
    Shape shape1 = new Square();
    Shape shape2 = new Square();
    
    Square square;
    Square square;

    if (shape1 instanceof Square square && square.getLength() > 0) {
    }
    if (shape2 instanceof Square square && square.getLength() > 0) {
    }
```

and we want to avoid that duplicate declaration, I had thoughts of using the pattern secretPatternvariable, but for each instanceof that is a new instance of that variable which means we end up with entries in the method locals.

What I could do is loop over those and try a different way of checking but that sounds like it defeats the purpose of the map and might be confusing, so I decided to introduce my own to avoid this kind of mixed logic. WDYT.

Reply:
I'm convinced (after you showing me live on a call), this probably needs to be this way - the other ways of making implicit locals don't really look like this use case, so we need something closer to what you've written than what already exists, thanks for explaining.